src/toucan/db.clj

```diff
@@ -152,6 +152,21 @@
   [& body]
   `(do-in-transaction (fn [] ~@body)))
 
+(defn do-in-transaction-with-rollback
+  "Execute F inside a DB transaction with rollback (will not commit changes on complete). Prefer macro form
+  `transaction-with-rollback` to using this directly."
+  [f]
+  (jdbc/with-db-transaction [conn (connection)]
+    (jdbc/db-set-rollback-only! conn)
```

Comment:
I think maybe this needs to be done before `jdbc/with-db-transaction`, because `jdbc/with-db-transaction` has special logic for handling connections where the connection is rollback only. Not sure about this tho

https://github.com/clojure/java.jdbc/blob/acffd9f5f216f8b8c1fc960c1d47b0b5feb56730/src/main/clojure/clojure/java/jdbc.clj#L808-L815

Reply:
It's because the [db-set-rollback-only!](https://github.com/clojure/java.jdbc/blob/2acbf2736a30634c9c75e6a1853e970ffa6bdbb4/src/main/clojure/clojure/java/jdbc.clj#L737) will set the outermost transaction as a rollback transaction. So it should be called inside the `with-db-transaction` wrapper.

I believe the tests I added should guarantee that behaviour.